### ISSUE TITLE
[feat] Local server proxy

### ISSUE THREAD

**Issue by raaymax** [tauri-apps/tauri#11953]
### Describe the problem

I would like to be able to "mount" `https://example.com/api/` under
`http://tauri.localhost/api/` so that all my requests could be same origin and easy to use. 
`@tauri-apps/plugin-http` is cool but it doesn't solve problem with urls in  `<img />` tags. 
Also cookie stores are separated so I can't access my private images if I log in using `plugin-http`.
Another problem with current approach is that request don't show up in devtools which makes debugging more difficult especially on android.

### Describe the solution you'd like

Basically I would need something like server proxy in Vite https://vite.dev/config/server-options#server-proxy
but for production use.

### Alternatives considered

alternatively ability to create own endpoints in rust. This could be even more versatile 
`POST http://tauri.localhost/api/custom/endpoint`

**Comment by raaymax** [comment]
I just discovered that I can use register_asynchronous_uri_scheme_protocol to replace the tauri protocol and add "endpoints" there. However, this approach requires me to implement asset serving as well. 🤔 There isn't much documentation on this topic.

Has anyone here tried something similar?

**Comment by raaymax** [comment]
Found this message on Discord: https://discord.com/channels/616186924390023171/1152560510907207700/1152567724304760892
But this doesn't work on android as it will be rewritten to http://proxy.localhost which is not the same origin :/

Something like this would be useful:
```rust
.register_uri_scheme_protocol("tauri".to_string(), |_ctx, request, responder| {
    if(request.uri().to_string().ends_with("/api/custom")){
        // do something else
    } else {
        // fallback to default tauri protocol
    });
})
````

**Comment by raaymax** [comment]
Another option would be to change  `http://${protocol}.localhost/${path}` into `http://localhost/${protocol}/${path}` for `windows` and `android`, but this would be breaking change

**Comment by gaojunran** [comment]
I'm quite confused about this, does Tauri allow the server proxy config in its front end `vite.config.ts`? I forward the url "/api" to the port 5100 of localhost(a python server, like what we often do in web development), but it only works in development but does not work in production. 

Maybe I should use tauri's plugin-http instead of axios or other fetch method from browser?

**Comment by chenqinggang001** [comment]
> Found this message on Discord: https://discord.com/channels/616186924390023171/1152560510907207700/1152567724304760892在 Discord 上找到这条消息：https://discord.com/channels/616186924390023171/1152560510907207700/1152567724304760892 But this doesn't work on android as it will be rewritten to http://proxy.localhost which is not the same origin :/但这在安卓上不行，因为它会被重写成 http://proxy.localhost，而这并不是同一个起源 ：/
> 
> Something like this would be useful:类似这样的内容会很有用：
> 
> .register_uri_scheme_protocol("tauri".to_string(), |_ctx, request, responder| {
>     if(request.uri().to_string().ends_with("/api/custom")){
>         // do something else
>     } else {
>         // fallback to default tauri protocol
>     });
> })

How can the production environment fallback to default tauri protocol, I tried `Some(uri.replace("tauri://localhost", "http://tauri.localhost/"))` but it wasn't feasible, @raaymax 

Because I wanted to create a same-origin proxy for the iframe, I used the 'tauri' protocol
.register_uri_scheme_protocol("tauri",...
```
let target = if uri. ends_with("api/custom") {
    Some(uri.replace("tauri://localhost/api/custom", "https://www.example.com))
} else {
   
};
```

**Comment by FabianLars** [comment]
i guess the idea would be to use the AssetResolver and serve the files directly like the tauri protocol does [here](https://github.com/tauri-apps/tauri/blob/afdd288eab7d8473fd6ad2796571a24c4873a8db/crates/tauri/src/protocol/tauri.rs). Not exactly straight forward 🤷

**Comment by chenqinggang001** [comment]
Thank you for your guidance. I'll give it a try

**Comment by chenqinggang001** [comment]
Do you know how to obtain front-end static resources in the production environment of Android @FabianLars

**Comment by FabianLars** [comment]
should work the same as on desktop (in build mode) via the AssetResolver eg https://github.com/tauri-apps/tauri/blob/afdd288eab7d8473fd6ad2796571a24c4873a8db/crates/tauri/src/protocol/tauri.rs#L208-L220

**Comment by chenqinggang001** [comment]
omg, you are my savior ~！ This difficult problem has been successfully solved
```
#[cfg(not(development))]
                {
                    let path_no_qf = uri.split(&['?', '#'][..]).next().unwrap_or("");
                    let path = path_no_qf
                        .strip_prefix("tauri://localhost")
                        .unwrap_or("")
                        .trim_start_matches('/');

                    if let Some(asset) = app_handle.asset_resolver().get(path.to_string()) {
                        let mut builder = tauri::http::Response::builder().status(200);
                        builder = builder.header("Access-Control-Allow-Origin", "*");
                        builder = builder.header("Content-Type", asset.mime_type);
                        if let Some(csp) = &asset.csp_header {
                            builder = builder.header("Content-Security-Policy", csp);
                        }

                        responder.respond(builder.body(asset.bytes).unwrap());
                    } else {
                        responder.respond(
                            tauri::http::Response::builder()
                                .status(404)
                                .body("Not Found (Asset)".as_bytes().to_vec())
                                .unwrap(),
                        );
                    }
                }
```
